sdk/go/pulumix/compose_test.go

```diff
@@ -0,0 +1,235 @@
+
+	ctx := context.Background()
+	result := pulumix.Compose(ctx, func(c *pulumix.Composer) (int, error) {
+		foo := pulumix.ComposeAwait(c, foo)
+		bar := pulumix.ComposeAwait(c, bar)
+		t.Errorf("Should not reach here, got: (%v, %v)", foo, bar)
+		return 0, nil // appease the compiler
+	})
+
+	_, _, _, _, err := pulumix.UnsafeAwait(ctx, result)
+	assert.Error(t, err)
+	assert.ErrorIs(t, err, giveErr)
+}
+
+func TestCompose_secret(t *testing.T) {
+	t.Parallel()
+
+	type User struct {
+		Username string
+		Password string
+	}
+
+	username := pulumix.Val("admin")
+	password := pulumi.ToSecret(pulumi.String("hunter2")).(pulumi.StringOutput)
+
+	ctx := context.Background()
+	result := pulumix.Compose(ctx, func(c *pulumix.Composer) (*User, error) {
+		return &User{
+			Username: pulumix.ComposeAwait(c, username),
+			Password: pulumix.ComposeAwait(c, password),
+		}, nil
+	})
+
+	got, known, secret, deps, err := pulumix.UnsafeAwait(ctx, result)
+	require.NoError(t, err)
+	assert.Equal(t, &User{Username: "admin", Password: "hunter2"}, got)
+	assert.True(t, known)
+	assert.True(t, secret)
+	assert.Empty(t, deps)
+}
+
+func TestCompose_unknown(t *testing.T) {
+	t.Parallel()
+
+	foo := pulumix.Val("foo")
+	bar := pulumix.Output[string]{
+		OutputState: internal.NewOutputState(nil, reflect.TypeOf("")),
+	}
+	internal.FulfillOutput(bar, nil, false /* known */, false, nil, nil)
+
+	ctx := context.Background()
+	result := pulumix.Compose(ctx, func(c *pulumix.Composer) (int, error) {
+		fooLen := len(pulumix.ComposeAwait(c, foo))
+		barLen := len(pulumix.ComposeAwait(c, bar))
+		t.Errorf("Should not reach here, got: (%v, %v)", fooLen, barLen)
+		return 0, nil // appease the compiler
+	})
+
+	_, known, _, _, err := pulumix.UnsafeAwait(ctx, result)
+	require.NoError(t, err)
+	assert.False(t, known)
+}
+
+func TestCompsoe_dependencies(t *testing.T) {
+	t.Parallel()
+
+	type Dependency struct {
+		internal.ResourceState
+
+		id int
+	}
+
+	dep1 := Dependency{id: 1}
+	dep2 := Dependency{id: 2}
+	dep3 := Dependency{id: 3}
+
+	a := pulumix.Output[int]{
+		OutputState: internal.NewOutputState(nil, reflect.TypeOf(0)),
+	}
+	internal.FulfillOutput(a, 1, true, false, []internal.Resource{dep1, dep2}, nil)
+
+	b := pulumix.Output[int]{
+		OutputState: internal.NewOutputState(nil, reflect.TypeOf(0)),
+	}
+	internal.FulfillOutput(b, 2, true, false, []internal.Resource{dep2, dep3}, nil)
+
+	ctx := context.Background()
+	result := pulumix.Compose(ctx, func(c *pulumix.Composer) (int, error) {
+		a := pulumix.ComposeAwait(c, a)
+		b := pulumix.ComposeAwait(c, b)
+		return a + b, nil
+	})
+
+	v, _, _, deps, err := pulumix.UnsafeAwait(ctx, result)
+	require.NoError(t, err)
+	assert.Equal(t, 3, v)
+
+	assert.ElementsMatch(t, []internal.Resource{dep1, dep2, dep3}, deps)
+}
+
+func TestCompose_panic(t *testing.T) {
+	t.Parallel()
+
+	ctx := context.Background()
+	result := pulumix.Compose(ctx, func(c *pulumix.Composer) (int, error) {
+		panic("great sadness")
```

Comment:
This is the test case for the behavior I was questioning earlier, do we want this?

Reply:
I could go either way. I _think_ we shouldn't catch the panic.

Reply:
I could go either way too, leaning to not catching the panic. See what other reviewers say!